### DISCUSSION TITLE
Disclosure width issues when placed in Flex

### DISCUSSION THREAD

**Discussion by hharnisc** [twilio-labs/paste#610] [General] [answered]
We're working on the NavigationItem component in the Sidebar and running into some issues with the Disclosure component. If we place the Disclosure component by itself it takes up the full space (as expected):

![image](https://user-images.githubusercontent.com/1388079/88706637-5c622e00-d0ce-11ea-8977-6d3d75a13044.png)

The markup looks like this:

```tsx

const NavigationItem = ({ item }: NavigationItemType) => {
  return (
    <Disclosure>
      <DisclosureHeading
        as="div"
        variant="heading50"
        marginBottom="space0"
        style={{ fontWeight: "normal" }}
      >
        <Flex vAlignContent="center">
          <ProductTwilioOrgIcon decorative={true} size="sizeIcon20" />
          <Text as="span" paddingLeft="space30">
            {item.title}
          </Text>
          {!item.tags
            ? null
            : item.tags.map((tag) => (
                <Text
                  key={tag}
                  as="span"
                  fontStyle="italic"
                  color="colorTextLink"
                  paddingLeft="space30"
                >
                  {tag}
                </Text>
              ))}
        </Flex>
      </DisclosureHeading>
    </Disclosure>
  );
};
```

The `NavigationItem` calls for a design that as a dropdown menu item to the right of the disclosure:

![image](https://user-images.githubusercontent.com/1388079/88707041-ed390980-d0ce-11ea-8d31-f1245dccbc82.png)

When I try to place another component next to the `Disclosure` with `Flex` the `Disclosure` shrinks to the size of the content:

![image](https://user-images.githubusercontent.com/1388079/88707206-2d988780-d0cf-11ea-9934-59ab9c4a9f6b.png)

(I used chrome dev tools to highlight the size of the parent `Flex` component)

Code looks like this:

```tsx
const NavigationItem = ({ item }: NavigationItemType) => {
  return (
    <Flex>
      <Disclosure>
        <DisclosureHeading
          as="div"
          variant="heading50"
          marginBottom="space0"
          style={{ fontWeight: "normal" }}
        >
          <Flex vAlignContent="center" grow>
            <ProductTwilioOrgIcon decorative={true} size="sizeIcon20" />
            <Text as="span" paddingLeft="space30">
              {item.title}
            </Text>
            {!item.tags
              ? null
              : item.tags.map((tag) => (
                  <Text
                    key={tag}
                    as="span"
                    fontStyle="italic"
                    color="colorTextLink"
                    paddingLeft="space30"
                  >
                    {tag}
                  </Text>
                ))}
          </Flex>
        </DisclosureHeading>
      </Disclosure>
    </Flex>
  );
};
```

After some investigation it looks like it *might* be due to the `Header` component as part of the `DisclosureHeader`: https://github.com/twilio-labs/paste/blob/main/packages/paste-core/components/disclosure/src/DisclosureHeading.tsx#L57-L61

If I use chrome dev tools to set `width="100%"` I can get the desired behavior:

![image](https://user-images.githubusercontent.com/1388079/88708007-435a7c80-d0d0-11ea-9acb-d74c5b4ac631.png)

I'm not sure what the best path forward would be, but a couple options might be to allow properties to be set on the `Heading` component separate from the `Box` in the `DisclosureHeading` or possibly just set the default `width="100%"` on the `Heading` if possible.

**Comment by hharnisc** [comment]
Digging a little deeper and adding `DisclosureContent` I think adding some level of customization to `DisclosureHeader` would solve our problem.

The following markup will solve our problem if we could set the width of the underlying `Header` component within `DisclosureHeader` to `100%`:

```tsx
const NavigationItem = ({ item, onSetPinnedProduct }: NavigationItemType) => {
  return (
    <Disclosure>
      <Flex>
        <Flex grow>
          <DisclosureHeading
            as="div"
            variant="heading50"
            marginBottom="space0"
            style={{ fontWeight: "normal" }}
          >
            <Flex vAlignContent="center">
              <ProductTwilioOrgIcon decorative={true} size="sizeIcon20" />
              <Text as="span" paddingLeft="space30">
                {item.title}
              </Text>
              {!item.tags
                ? null
                : item.tags.map((tag) => (
                    <Text
                      key={tag}
                      as="span"
                      fontStyle="italic"
                      color="colorTextLink"
                      paddingLeft="space30"
                    >
                      {tag}
                    </Text>
                  ))}
            </Flex>
          </DisclosureHeading>
        </Flex>
        <NavigationItemDropdown
          product={item}
          onSetPinnedProduct={onSetPinnedProduct}
        />
      </Flex>
      <DisclosureContent>
        {item.productPages.map((productPage) => (
          <NavigationSubItem
            productKey={item.key}
            subItem={productPage}
            key={productPage.key}
          />
        ))}
      </DisclosureContent>
    </Disclosure>
  );
};
```

**Reply by hharnisc** [reply]
Relevant Console changes: https://code.hq.twilio.com/twilio/console/pull/40

**Comment by kate-porter** [comment, marked as answer]
Hey @hharnisc - I'm sharing this with the eng team, and we'll followup with you in a bit. Thanks!

**Reply by hharnisc** [reply]
Thank you @katieporter 🙌 @SiTaggart also requested I create a self contained example in CodeSandbox. Here's a link: https://codesandbox.io/s/ecstatic-grothendieck-ztrsf?file=/src/index.tsx

Focusing on the `Disclosure` reveals the issue

**Reply by richbachman** [reply]
Hi @hharnisc. This may be an issue with us over-rotating on the styles in Flex. We're going to take a look and evaluate that more. 

In the meantime, you should be to accomplish what you're looking for with Box. I've forked your sandbox and replaced the Flex components with Box: https://codesandbox.io/s/new-night-udlwz?file=/src/index.tsx

Let us know if we're missing anything.

**Reply by hharnisc** [reply]
Thank you @richbachman I'll give this a try and report back, but with the example it looks it should work. Thank you 🙌

**Reply by hharnisc** [reply]
To close the loop this resolved the issue -- will keep an eye out for other places where we've used `Flex`. Thanks again @richbachman  and @katieporter

**Reply by kate-porter** [reply]
Thanks @hharnisc!

**Comment by kate-porter** [comment]
Thanks for reproducing that in the sandbox @hharnisc! Super helpful.

Sounds like the recommendation would be to use the [Disclosure Primitive](https://paste.twilio.design/primitives/disclosure-primitive/) in this case—rather than the Disclosure Component—since this use case will require custom styling. I think @SiTaggart can jump in and put together a sandbox showing how that would work.
